query-builder/src/main/java/com/datastax/oss/driver/internal/querybuilder/select/DefaultSelect.java

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.querybuilder.select;
+
+import com.datastax.oss.driver.api.core.CqlIdentifier;
+import com.datastax.oss.driver.api.core.cql.SimpleStatement;
+import com.datastax.oss.driver.api.querybuilder.BindMarker;
+import com.datastax.oss.driver.api.querybuilder.relation.Relation;
+import com.datastax.oss.driver.api.querybuilder.select.Select;
+import com.datastax.oss.driver.api.querybuilder.select.SelectFrom;
+import com.datastax.oss.driver.api.querybuilder.select.Selector;
+import com.datastax.oss.driver.internal.core.metadata.schema.ScriptBuilder;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.util.function.Function;
+
+public class DefaultSelect implements SelectFrom, Select {
+
+  private static final ImmutableList<Selector> SELECT_ALL = ImmutableList.of(AllSelector.INSTANCE);
+
+  private final CqlIdentifier keyspace;
+  private final CqlIdentifier table;
+  private final boolean isJson;
+  private final boolean isDistinct;
+  private final ImmutableList<Selector> selectors;
+  private final ImmutableList<Relation> relations;
+  private final Object limit;
+  private final boolean allowsFiltering;
+
+  public DefaultSelect(CqlIdentifier keyspace, CqlIdentifier table) {
+    this(keyspace, table, false, false, ImmutableList.of(), ImmutableList.of(), null, false);
+  }
+
+  /**
+   * This constructor is public only as a convenience for custom extensions of the query builder.
+   *
+   * @param selectors if it contains {@link AllSelector#INSTANCE}, that must be the only element.
+   *     This isn't re-checked because methods that call this constructor internally already do it,
+   *     make sure you do it yourself.
+   */
+  public DefaultSelect(
```

Comment:
I'm being very liberal with visibility in internal packages. The query builder is used a lot in 3rd-party frameworks, and it's nice to keep the door open if people want means to clone statements (see comment on the JIRA ticket).
Similarly, all the fields are exposed with public getters for introspection.